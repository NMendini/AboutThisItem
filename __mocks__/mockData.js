module.exports = {
  details: {
    "highlights": [
        "MAGNI REPELLAT AUT: Facere odio fugiat nesciunt voluptatem rerum omnis accusantium aut adipisci.",
        "ATQUE CONSEQUATUR ET: Consequatur error fugit sint possimus.\nVel id cum dolor.",
        "IURE AT SOLUTA: Porro dolore libero nisi.\nVel rerum non aliquid ut.",
        "QUOS EUM VEL: Aspernatur excepturi qui eaque consequatur voluptatem id consectetur officia dicta.\nSit aut aliquid sunt dignissimos id est perferendis deleniti deleniti.\nFuga vel quo qui odit animi et id.\nQuos rerum corporis quis pariatur nulla sed id.\nUt consectetur temporibus.",
        "EAQUE VOLUPTATEM PARIATUR: Sit repellat rem veritatis nihil quaerat eum quia quis.",
        "ATQUE EA AUT: Corporis quas magnam adipisci consequatur dolore et ipsum et quae.\nSed hic omnis ducimus rerum odit possimus.\nTempore adipisci eos quod rem.\nEa quos architecto corporis.\nDolore in rem quisquam ut quia."
    ],
    "specifications": {
        "Item Number": "1",
        "Frozen": "aut dicta eligendi",
        "Rubber": "et odit voluptas",
        "Fresh": "quas qui iste",
        "Metal": "nostrum officia sunt",
        "Wooden": "dolor fugit nihil",
        "Granite": "enim architecto at",
        "Cotton": "sint repellendus velit",
        "Plastic": "omnis ratione iure",
        "Steel": "et quisquam animi",
        "Height": "17.25 inches",
        "Width": "24.07 inches",
        "Length": "27.23 inches",
        "Weight": "5.48 pounds"
    },
    "description": "Ergonomic executive chair upholstered in bonded black leather and PVC padded seat and back for all-day comfort and support. Labore quasi praesentium reiciendis et expedita magnam ut. Aut ex quod voluptatem dolorem id et aut fuga et. Sit distinctio magni sit suscipit molestias quae et. Vel quo fugiat blanditiis placeat. Consequuntur itaque aliquam sequi ut.\n \rIusto quas corrupti tempora. Animi neque neque ut blanditiis aliquam eveniet rem. Sed officia reiciendis nulla vel aliquam eum iste tenetur.\n \rQuia debitis libero rem ad qui officia alias ut optio. Vel non qui iure. Nobis doloribus voluptatem quisquam omnis. Et rerum quae sint atque laborum tenetur excepturi ullam. Officiis et aut ea."
  },
  questions: [
    {
        "question_id": 1,
        "user_name": "Dedrick",
        "question": "Aut eum aut laboriosam id laborum vel omnis voluptas?",
        "created_at": "2020-04-02T17:51:57.000Z",
        "product_id": 91
    },
    {
        "question_id": 2,
        "user_name": "Freda",
        "question": "Ea quisquam atque molestiae non qui accusantium?",
        "created_at": "2020-05-18T20:14:35.000Z",
        "product_id": 91
    }
  ],
  answers: [
    {
        "answer_id": 1,
        "user_name": "Nichole",
        "question_id": 1,
        "answer": "Neque officiis quod quo perferendis voluptatum aut sed. Aut in pariatur tempora et animi saepe non. Iure neque quia qui qui qui excepturi. Quod vitae pariatur nisi neque quidem voluptates maxime qui.",
        "created_at": "2020-01-15T19:45:38.000Z",
        "helpful": 1,
        "not_helpful": 12,
        "product_id": 1
    },
    {
        "answer_id": 2,
        "user_name": "Arianna",
        "question_id": 1,
        "answer": "Qui sit vel aut molestiae. Expedita quia ipsam. Soluta accusantium aspernatur consequatur quibusdam magnam quas. Voluptas fuga sapiente nam distinctio deserunt fuga atque.",
        "created_at": "2020-05-03T22:34:22.000Z",
        "helpful": 19,
        "not_helpful": 27,
        "product_id": 1
    },
    {
      "answer_id": 3,
      "user_name": "Joshua",
      "question_id": 2,
      "answer": "Soluta optio deserunt ipsam aut hic nostrum quaerat. Voluptatem accusamus harum.",
      "created_at": "2020-03-09T08:11:04.000Z",
      "helpful": 16,
      "not_helpful": 35,
      "product_id": 1
    },
    {
        "answer_id": 4,
        "user_name": "Hubert",
        "question_id": 2,
        "answer": "Nisi distinctio praesentium. Qui maxime qui a laborum eum quia sed vero quae. Cupiditate voluptatem aut culpa dolores omnis enim voluptas recusandae. Assumenda et voluptatem sed unde impedit autem cupiditate cum.",
        "created_at": "2020-03-12T15:16:48.000Z",
        "helpful": 37,
        "not_helpful": 39,
        "product_id": 1
    }
  ]
}