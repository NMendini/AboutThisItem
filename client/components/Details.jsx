import React from 'react';
import HighlightsList from './HighlightsList.jsx';
import SpecificationsList from './SpecificationsList.jsx';
import styles from '../assets/style.css';

class Details extends React.Component {
  constructor(props) {
    super(props);
    this.handleToggle = this.handleToggle.bind(this);

    this.btnExpand = React.createRef();
    this.specsContainer = React.createRef();
    this.detailsContainer = React.createRef();

    this.state = {
      toggled: false
    };
  }

  handleToggle() {
    this.setState({
      toggled: !this.state.toggled
    }, () => {
      this.props.toggle(this.btnExpand.current);
      if (this.state.toggled) {
        this.specsContainer.current.classList.remove(`${styles.collapsed}`);
      } else {
        this.specsContainer.current.classList.add(`${styles.collapsed}`);
      }
    });
  }

  render() {

    return (
      <div id="details-container" className={this.props.visible} ref={this.detailsContainer}>
        <div className={styles.containerWhite}>
          <div>
            <div className={styles.marginLeft}>
              <h3 className="highlights">Highlights</h3>
              <HighlightsList highlights={this.props.highlights}/>
            </div>
          </div>
        </div>
        <div id="specs-container" ref={this.specsContainer}className={`${styles.containerWhite} ${styles.collapsed}`}>
          <div>
            <div className={styles.marginLeft}>
              <div className={styles.containerSpecDesc}>
                <div className={styles.column}>
                  <h3 className="Specifications">Specifications</h3>
                  <SpecificationsList specifications={this.props.specifications}/>
                </div>
                <div className={styles.column}>
                  <h3 className="Description">Description</h3>
                  <div className={styles.descriptionContainer}>{this.props.description}</div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <button type="button" ref={this.btnExpand} id="btn-expand" className={styles.btn} onClick={this.handleToggle}>Show more</button>
      </div>
    );
  }
}

export default Details;