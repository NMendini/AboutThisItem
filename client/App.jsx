import 'babel-polyfill';
import React from 'react';
import ReactDOM from 'react-dom';
import axios from 'axios';
import QAndA from './components/QAndA.jsx';
import ShippingAndReturns from './components/ShippingAndReturns.jsx';
import Details from './components/Details.jsx';
import styles from './assets/style.css';


class App extends React.Component {
  constructor(props) {
    super(props);
    this.handleClick = this.handleClick.bind(this);

    this.state = {
      selected: 'details',
      showToggle: false,
      highlights: [],
      description: '',
      specifications: {},
      questions: [],
      questionCount: 0,
      answers: [],
      product: 1
    };

    // INITIAL API CALLS FOR DATA
  }
  componentDidMount() {
    this.getProductDetails();
    this.getQuestions();
  }

  // API CALLS
  getProductDetails() {
    axios({
      method: 'GET',
      url: `/api/products/${this.state.product}/details`
    })
      .then(result => {
        this.setState({
          description: result.data.description,
          highlights: result.data.highlights,
          specifications: result.data.specifications
        });
      }, (err) => {
        console.error(err);
      });
  }

  getQuestions() {
    axios({
      method: 'GET',
      url: `/api/products/${this.state.product}/questions`,
    })
      .then(results => {
        this.setState({
          questionCount: results.data.length,
          questions: results.data
        });
      }, (err) => {
        console.error(err);
      });
  }

  getAnswers() {
    axios({
      method: 'GET',
      url: `/api/products/${this.state.product}/answers`
    })
      .then(results => {
        this.setState({
          answers: results.data
        });
      }, (err) => {
        console.error(err);
      });
  }

  // HANDLERS
  handleClick(e) {
    document.getElementById(this.state.selected).classList.remove(`${styles.selected}`);

    document.getElementById('shipping-returns-container').classList.add(`${styles.displayNone}`);
    document.getElementById('details-container').classList.add(`${styles.displayNone}`);
    document.getElementById('q-and-a-container').classList.add(`${styles.displayNone}`);

    this.setState({
      selected: e.target.id
    }, ()=> {
      document.getElementById(this.state.selected).classList.add(`${styles.selected}`);
      if (this.state.selected === 'shipping') {
        document.getElementById('shipping-returns-container').classList.remove(`${styles.displayNone}`);
      }
      if (this.state.selected === 'details') {
        document.getElementById('details-container').classList.remove(`${styles.displayNone}`);
      }
      if (this.state.selected === 'q-and-a') {
        document.getElementById('q-and-a-container').classList.remove(`${styles.displayNone}`);
        this.getAnswers();
      }
    });
  }

  handleShowToggle() {
    var button = document.getElementById('btn expand');

    if (!this.state.showToggle) {
      return this.setState({
        showToggle: !this.state.showToggle
      }, () => {
        button.innerHTML = 'Show less';
      });
    }
    this.setState({
      showToggle: false
    }, () => {
      button.innerHTML = 'Show more';
    });
  }

  render() {
    return (
      <div id={styles.mainContainer}>
        <h2 id={styles.moduleTitle}>About this item</h2>
        <div id={styles.tabHeading}>
          <div className={styles.marginLeft}>
            <ul id={styles.tabList}>
              <li id="details" className={`${styles.tab} ${styles.details} ${styles.selected}`} onClick={this.handleClick}>Details</li>
              <li id="shipping" className={`${styles.tab} shipping`} onClick={this.handleClick}>Shipping & Returns</li>
              <li id="q-and-a" className={`${styles.tab} q-and-a`} onClick={this.handleClick}>Q&A ({this.state.questionCount})</li>
            </ul>
          </div>
        </div>
        <Details
          toggle={this.handleShowToggle.bind(this)}
          highlights={this.state.highlights}
          toggleStatus={this.state.showToggle}
          description={this.state.description}
          specifications={this.state.specifications}
        />
        <ShippingAndReturns specs={this.state.specifications}/>
        <QAndA
          questions={this.state.questions}
          answers={this.state.answers}
          questionCount={this.state.questionCount}
        />

      </div>
    );
  }
}

export default App;