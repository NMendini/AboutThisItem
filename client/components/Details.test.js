/**
 * @jest-environment jsdom
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';
import mockAxios from 'jest-mock-axios';
import { details, questions, answers }from '../../__mocks__/mockData.js';


// COMPONENTS
import App from '../App.jsx';
import Details from './Details.jsx';

let component;
let mockToggle;

beforeAll(async () => {
  mockToggle = jest.fn();
  Details.prototype.handleToggle = mockToggle;
  component = await shallow(<Details />);

});

afterAll(async () => {
  await component.unmount();
});

describe('Details Component', () => {
  it('should render Details component correctly', async () => {
      expect(component).toMatchSnapshot();
  });

  it('should call handleToggle function on button click', async () => {
    component.find('button').simulate('click');
    expect(mockToggle).toHaveBeenCalled();
  });
});