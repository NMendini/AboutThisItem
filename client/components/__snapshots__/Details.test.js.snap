// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details Component should render Details component correctly 1`] = `
<div
  id="details-container"
>
  <div
    className="containerWhite"
  >
    <div>
      <div
        className="marginLeft"
      >
        <h3
          className="highlights"
        >
          Highlights
        </h3>
        <HighlightsList />
      </div>
    </div>
  </div>
  <div
    className="containerWhite collapsed"
    id="specs-container"
  >
    <div>
      <div
        className="marginLeft"
      >
        <div
          className="containerSpecDesc"
        >
          <div
            className="column"
          >
            <h3
              className="Specifications"
            >
              Specifications
            </h3>
            <SpecificationsList />
          </div>
          <div
            className="column"
          >
            <h3
              className="Description"
            >
              Description
            </h3>
            <div
              className="descriptionContainer"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <button
    className="btn"
    id="btn-expand"
    onClick={[Function]}
    type="button"
  >
    Show more
  </button>
</div>
`;
